File: pregnancy_diary.go
// pregnancy_diary.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"strconv"
	"time"
)

const dataFile = "pregnancy.json"

type Entry struct {
	Date     string `json:"date"`
	Week     int    `json:"week"`
	Note     string `json:"note"`
	Symptoms string `json:"symptoms"`
	Created  string `json:"created"`
}

type Reminder struct {
	ID      int    `json:"id"`
	Date    string `json:"date"`
	Note    string `json:"note"`
	Done    bool   `json:"done"`
	Created string `json:"created"`
}

type DiaryData struct {
	StartDate string     `json:"start_date"`
	Entries   []Entry    `json:"entries"`
	Reminders []Reminder `json:"reminders"`
}

type PregnancyDiary struct {
	data           DiaryData
	nextReminderID int
}

func NewPregnancyDiary() *PregnancyDiary {
	d := &PregnancyDiary{data: DiaryData{Entries: []Entry{}, Reminders: []Reminder{}}}
	d.load()
	return d
}

func (d *PregnancyDiary) load() {
	file, err := os.ReadFile(dataFile)
	if err != nil {
		return
	}
	if err := json.Unmarshal(file, &d.data); err != nil {
		return
	}
	maxID := 0
	for _, r := range d.data.Reminders {
		if r.ID > maxID {
			maxID = r.ID
		}
	}
	d.nextReminderID = maxID + 1
}

func (d *PregnancyDiary) save() {
	data, _ := json.MarshalIndent(d.data, "", "  ")
	os.WriteFile(dataFile, data, 0644)
}

func (d *PregnancyDiary) setStartDate(dateStr string) {
	d.data.StartDate = dateStr
	d.save()
	fmt.Printf("\033[32mДата начала беременности установлена: %s\033[0m\n", dateStr)
}

func (d *PregnancyDiary) calculateWeek(dateStr string) *int {
	if d.data.StartDate == "" {
		return nil
	}
	start, _ := time.Parse("2006-01-02", d.data.StartDate)
	target, _ := time.Parse("2006-01-02", dateStr)
	delta := int(target.Sub(start).Hours() / 24)
	if delta < 0 {
		return nil
	}
	week := delta/7 + 1
	return &week
}

func (d *PregnancyDiary) addEntry(dateStr string, week int, note, symptoms string) {
	if week == 0 {
		w := d.calculateWeek(dateStr)
		if w == nil {
			fmt.Println("\033[31mНе удалось определить неделю. Укажите --week или --start-date\033[0m")
			return
		}
		week = *w
	}
	entry := Entry{
		Date:     dateStr,
		Week:     week,
		Note:     note,
		Symptoms: symptoms,
		Created:  time.Now().Format(time.RFC3339),
	}
	d.data.Entries = append(d.data.Entries, entry)
	d.save()
	fmt.Printf("\033[32mЗапись добавлена: %s, неделя %d\033[0m\n", dateStr, week)
}

func (d *PregnancyDiary) addReminder(dateStr, note string) {
	reminder := Reminder{
		ID:      d.nextReminderID,
		Date:    dateStr,
		Note:    note,
		Done:    false,
		Created: time.Now().Format(time.RFC3339),
	}
	d.nextReminderID++
	d.data.Reminders = append(d.data.Reminders, reminder)
	d.save()
	fmt.Printf("\033[32mНапоминание добавлено (ID: %d)\033[0m\n", reminder.ID)
}

func (d *PregnancyDiary) listEntries() {
	if len(d.data.Entries) == 0 {
		fmt.Println("\033[33mНет записей.\033[0m")
		return
	}
	fmt.Println("\033[36m📖 Дневник беременности:\033[0m")
	for _, e := range d.data.Entries {
		fmt.Printf("  %s | Неделя %d | %s | Симптомы: %s\n", e.Date, e.Week, e.Note, e.Symptoms)
	}
}

func (d *PregnancyDiary) listReminders() {
	if len(d.data.Reminders) == 0 {
		fmt.Println("\033[33mНет напоминаний.\033[0m")
		return
	}
	fmt.Println("\033[36m⏰ Напоминания:\033[0m")
	for _, r := range d.data.Reminders {
		status := "\033[32m✓\033[0m"
		if !r.Done {
			status = "\033[31m✗\033[0m"
		}
		fmt.Printf("  %d. %s %s - %s\n", r.ID, status, r.Date, r.Note)
	}
}

func (d *PregnancyDiary) markReminderDone(id int) {
	for i, r := range d.data.Reminders {
		if r.ID == id {
			d.data.Reminders[i].Done = true
			d.save()
			fmt.Printf("\033[32mНапоминание #%d отмечено как выполненное.\033[0m\n", id)
			return
		}
	}
	fmt.Printf("\033[31mНапоминание #%d не найдено.\033[0m\n", id)
}

func (d *PregnancyDiary) exportJSON(filename string) {
	data, _ := json.MarshalIndent(d.data, "", "  ")
	os.WriteFile(filename, data, 0644)
	fmt.Printf("\033[32mЭкспортировано в %s (JSON)\033[0m\n", filename)
}

func (d *PregnancyDiary) exportCSV(filename string) {
	f, _ := os.Create(filename)
	defer f.Close()
	w := csv.NewWriter(f)
	defer w.Flush()
	w.Write([]string{"date", "week", "note", "symptoms", "created"})
	for _, e := range d.data.Entries {
		w.Write([]string{e.Date, strconv.Itoa(e.Week), e.Note, e.Symptoms, e.Created})
	}
	fmt.Printf("\033[32mЭкспортировано в %s (CSV)\033[0m\n", filename)
}

func (d *PregnancyDiary) exportTXT(filename string) {
	f, _ := os.Create(filename)
	defer f.Close()
	fmt.Fprintf(f, "Дата начала беременности: %s\n\n", d.data.StartDate)
	fmt.Fprintln(f, "=== ЗАПИСИ ===")
	for _, e := range d.data.Entries {
		fmt.Fprintf(f, "%s | Неделя %d | %s | Симптомы: %s\n", e.Date, e.Week, e.Note, e.Symptoms)
	}
	fmt.Fprintln(f, "\n=== НАПОМИНАНИЯ ===")
	for _, r := range d.data.Reminders {
		done := "[ ]"
		if r.Done {
			done = "[x]"
		}
		fmt.Fprintf(f, "%d. %s - %s %s\n", r.ID, r.Date, r.Note, done)
	}
	fmt.Printf("\033[32mЭкспортировано в %s (TXT)\033[0m\n", filename)
}

func main() {
	var (
		add        bool
		dateStr    string
		week       int
		note       string
		symptoms   string
		startDate  string
		remind     bool
		list       bool
		reminders  bool
		remindDone int
		expJson    string
		expCsv     string
		expTxt     string
	)
	flag.BoolVar(&add, "add", false, "Добавить запись")
	flag.StringVar(&dateStr, "date", "", "Дата (YYYY-MM-DD)")
	flag.IntVar(&week, "week", 0, "Неделя беременности")
	flag.StringVar(&note, "note", "", "Заметка")
	flag.StringVar(&symptoms, "symptoms", "", "Симптомы")
	flag.StringVar(&startDate, "start-date", "", "Дата начала беременности")
	flag.BoolVar(&remind, "remind", false, "Добавить напоминание")
	flag.BoolVar(&list, "list", false, "Показать записи")
	flag.BoolVar(&reminders, "reminders", false, "Показать напоминания")
	flag.IntVar(&remindDone, "remind-done", 0, "Отметить напоминание как выполненное")
	flag.StringVar(&expJson, "export-json", "", "Экспорт в JSON")
	flag.StringVar(&expCsv, "export-csv", "", "Экспорт в CSV")
	flag.StringVar(&expTxt, "export-txt", "", "Экспорт в TXT")
	flag.Parse()

	diary := NewPregnancyDiary()

	if startDate != "" {
		diary.setStartDate(startDate)
	} else if add {
		if dateStr == "" {
			fmt.Println("\033[31mДля добавления записи требуется --date\033[0m")
			os.Exit(1)
		}
		diary.addEntry(dateStr, week, note, symptoms)
	} else if remind {
		if dateStr == "" || note == "" {
			fmt.Println("\033[31mДля добавления напоминания требуются --date и --note\033[0m")
			os.Exit(1)
		}
		diary.addReminder(dateStr, note)
	} else if list {
		diary.listEntries()
	} else if reminders {
		diary.listReminders()
	} else if remindDone != 0 {
		diary.markReminderDone(remindDone)
	} else if expJson != "" {
		diary.exportJSON(expJson)
	} else if expCsv != "" {
		diary.exportCSV(expCsv)
	} else if expTxt != "" {
		diary.exportTXT(expTxt)
	} else {
		fmt.Println("Используйте --help для справки.")
	}
}

File: pregnancy_diary.rs
// pregnancy_diary.rs
use chrono::{DateTime, Local, NaiveDate, Utc};
use clap::{App, Arg};
use serde::{Deserialize, Serialize};
use serde_json;
use std::fs;
use std::io::Write;
use colored::*;

const DATA_FILE: &str = "pregnancy.json";

#[derive(Serialize, Deserialize)]
struct Entry {
    date: String,
    week: u32,
    note: String,
    symptoms: String,
    created: String,
}

#[derive(Serialize, Deserialize)]
struct Reminder {
    id: u32,
    date: String,
    note: String,
    done: bool,
    created: String,
}

#[derive(Serialize, Deserialize)]
struct DiaryData {
    start_date: String,
    entries: Vec<Entry>,
    reminders: Vec<Reminder>,
}

struct PregnancyDiary {
    data: DiaryData,
    next_reminder_id: u32,
}

impl PregnancyDiary {
    fn new() -> Self {
        let mut d = PregnancyDiary {
            data: DiaryData {
                start_date: "".to_string(),
                entries: Vec::new(),
                reminders: Vec::new(),
            },
            next_reminder_id: 1,
        };
        d.load();
        d
    }

    fn load(&mut self) {
        if let Ok(data) = fs::read_to_string(DATA_FILE) {
            if let Ok(mut diary_data) = serde_json::from_str::<DiaryData>(&data) {
                self.data = diary_data;
                self.next_reminder_id = self.data.reminders.iter().map(|r| r.id).max().unwrap_or(0) + 1;
            }
        }
    }

    fn save(&self) {
        let json = serde_json::to_string_pretty(&self.data).unwrap();
        fs::write(DATA_FILE, json).unwrap();
    }

    fn set_start_date(&mut self, date_str: &str) {
        self.data.start_date = date_str.to_string();
        self.save();
        println!("{}", format!("Дата начала беременности установлена: {}", date_str).green());
    }

    fn calculate_week(&self, date_str: &str) -> Option<u32> {
        if self.data.start_date.is_empty() {
            return None;
        }
        let start = NaiveDate::parse_from_str(&self.data.start_date, "%Y-%m-%d").ok()?;
        let target = NaiveDate::parse_from_str(date_str, "%Y-%m-%d").ok()?;
        let delta = (target - start).num_days();
        if delta < 0 { return None; }
        Some((delta / 7) as u32 + 1)
    }

    fn add_entry(&mut self, date_str: &str, week: Option<u32>, note: &str, symptoms: &str) {
        let week = if let Some(w) = week {
            w
        } else {
            match self.calculate_week(date_str) {
                Some(w) => w,
                None => {
                    println!("{}", "Не удалось определить неделю. Укажите --week или --start-date".red());
                    return;
                }
            }
        };
        let entry = Entry {
            date: date_str.to_string(),
            week,
            note: note.to_string(),
            symptoms: symptoms.to_string(),
            created: Utc::now().to_rfc3339(),
        };
        self.data.entries.push(entry);
        self.save();
        println!("{}", format!("Запись добавлена: {}, неделя {}", date_str, week).green());
    }

    fn add_reminder(&mut self, date_str: &str, note: &str) {
        let reminder = Reminder {
            id: self.next_reminder_id,
            date: date_str.to_string(),
            note: note.to_string(),
            done: false,
            created: Utc::now().to_rfc3339(),
        };
        self.next_reminder_id += 1;
        self.data.reminders.push(reminder);
        self.save();
        println!("{}", format!("Напоминание добавлено (ID: {})", reminder.id).green());
    }

    fn list_entries(&self) {
        if self.data.entries.is_empty() {
            println!("{}", "Нет записей.".yellow());
            return;
        }
        println!("{}", "📖 Дневник беременности:".cyan());
        for e in &self.data.entries {
            println!("  {} | Неделя {} | {} | Симптомы: {}", e.date, e.week, e.note, e.symptoms);
        }
    }

    fn list_reminders(&self) {
        if self.data.reminders.is_empty() {
            println!("{}", "Нет напоминаний.".yellow());
            return;
        }
        println!("{}", "⏰ Напоминания:".cyan());
        for r in &self.data.reminders {
            let status = if r.done { "✓".green() } else { "✗".red() };
            println!("  {}. {} {} - {}", r.id, status, r.date, r.note);
        }
    }

    fn mark_reminder_done(&mut self, id: u32) {
        if let Some(r) = self.data.reminders.iter_mut().find(|r| r.id == id) {
            r.done = true;
            self.save();
            println!("{}", format!("Напоминание #{} отмечено как выполненное.", id).green());
        } else {
            println!("{}", format!("Напоминание #{} не найдено.", id).red());
        }
    }

    fn export_json(&self, filename: &str) {
        let json = serde_json::to_string_pretty(&self.data).unwrap();
        fs::write(filename, json).unwrap();
        println!("{}", format!("Экспортировано в {} (JSON)", filename).green());
    }

    fn export_csv(&self, filename: &str) {
        let mut wtr = csv::Writer::from_path(filename).unwrap();
        wtr.write_record(&["date", "week", "note", "symptoms", "created"]).unwrap();
        for e in &self.data.entries {
            wtr.write_record(&[&e.date, &e.week.to_string(), &e.note, &e.symptoms, &e.created]).unwrap();
        }
        wtr.flush().unwrap();
        println!("{}", format!("Экспортировано в {} (CSV)", filename).green());
    }

    fn export_txt(&self, filename: &str) {
        let mut content = String::new();
        content.push_str(&format!("Дата начала беременности: {}\n\n", self.data.start_date));
        content.push_str("=== ЗАПИСИ ===\n");
        for e in &self.data.entries {
            content.push_str(&format!("{} | Неделя {} | {} | Симптомы: {}\n", e.date, e.week, e.note, e.symptoms));
        }
        content.push_str("\n=== НАПОМИНАНИЯ ===\n");
        for r in &self.data.reminders {
            let done = if r.done { "[x]" } else { "[ ]" };
            content.push_str(&format!("{}. {} - {} {}\n", r.id, r.date, r.note, done));
        }
        fs::write(filename, content).unwrap();
        println!("{}", format!("Экспортировано в {} (TXT)", filename).green());
    }
}

fn main() {
    let matches = App::new("Pregnancy Diary")
        .arg(Arg::with_name("add").long("add").help("Добавить запись"))
        .arg(Arg::with_name("date").long("date").takes_value(true).help("Дата (YYYY-MM-DD)"))
        .arg(Arg::with_name("week").long("week").takes_value(true).help("Неделя беременности"))
        .arg(Arg::with_name("note").long("note").takes_value(true).help("Заметка"))
        .arg(Arg::with_name("symptoms").long("symptoms").takes_value(true).help("Симптомы"))
        .arg(Arg::with_name("start-date").long("start-date").takes_value(true).help("Дата начала беременности"))
        .arg(Arg::with_name("remind").long("remind").help("Добавить напоминание"))
        .arg(Arg::with_name("list").long("list").help("Показать записи"))
        .arg(Arg::with_name("reminders").long("reminders").help("Показать напоминания"))
        .arg(Arg::with_name("remind-done").long("remind-done").takes_value(true).help("Отметить напоминание как выполненное"))
        .arg(Arg::with_name("export-json").long("export-json").takes_value(true).help("Экспорт в JSON"))
        .arg(Arg::with_name("export-csv").long("export-csv").takes_value(true).help("Экспорт в CSV"))
        .arg(Arg::with_name("export-txt").long("export-txt").takes_value(true).help("Экспорт в TXT"))
        .get_matches();

    let mut diary = PregnancyDiary::new();

    if let Some(date) = matches.value_of("start-date") {
        diary.set_start_date(date);
    } else if matches.is_present("add") {
        let date = matches.value_of("date").expect("--date required");
        let week = matches.value_of("week").map(|s| s.parse().unwrap());
        let note = matches.value_of("note").unwrap_or("");
        let symptoms = matches.value_of("symptoms").unwrap_or("");
        diary.add_entry(date, week, note, symptoms);
    } else if matches.is_present("remind") {
        let date = matches.value_of("date").expect("--date required");
        let note = matches.value_of("note").expect("--note required");
        diary.add_reminder(date, note);
    } else if matches.is_present("list") {
        diary.list_entries();
    } else if matches.is_present("reminders") {
        diary.list_reminders();
    } else if let Some(id_str) = matches.value_of("remind-done") {
        let id: u32 = id_str.parse().expect("Invalid ID");
        diary.mark_reminder_done(id);
    } else if let Some(file) = matches.value_of("export-json") {
        diary.export_json(file);
    } else if let Some(file) = matches.value_of("export-csv") {
        diary.export_csv(file);
    } else if let Some(file) = matches.value_of("export-txt") {
        diary.export_txt(file);
    } else {
        println!("Используйте --help для справки.");
    }
}
